### PR TITLE
DS-714 | @ericbakenhus | Search bar with autocomplete

### PR DESCRIPTION
**Pull request by ericbakenhus**
# READY FOR REVIEW

# Summary
- Add search bar with autocomplete (mobile search will be handled in DS-716)

# Review By (Date)
- Retro

# Criticality
- Low

# Review Tasks

## Setup tasks and/or behavior to test

1. Check out this branch
2. Navigate to /events/discovery
3. Verify the search bar allows for searching by title
4. Verify suggestions are shown below the input after a value is typed
5. Verify the search button works
6. Verify a chip is shown when a search query is active
7. Verify a clear button is shown below the search input when a query is active

# Associated Issues and/or People
- DS-714

### CODE REVIEW

File: src/components/events-discovery/components/Chips/ChipsComponent.jsx [hunk 1]
```diff
@@ -64,14 +65,30 @@ export const ChipsComponent = () => {
...
+        attribute: 'query',
+        label: query,
+        remove: () => refine(''),
+      },
+    ];
+  }, [processedItems, query, refine]);
+
+  if (!canRefine && !query) {
     return null;
   }
```
Comment (ericbakenhus) [inline src/components/events-discovery/components/Chips/ChipsComponent.jsx:86]: Add query as a chip.

File: src/components/events-discovery/components/SearchBar/SearchBar.jsx
Comment (ericbakenhus) [inline src/components/events-discovery/components/SearchBar/SearchBar.jsx:1]: Search bar component.

File: src/components/events-discovery/components/SearchBar/SearchBar.jsx [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+  const { query, refine } = useSearchBox();
+  const [value, setValue] = useState(query);
+  const [inputValue, setInputValue] = useState(query);
+  const debouncedInputValue = useDebouncedValue(inputValue);
+  const [options, setOptions] = useState([]);
+
+  useEffect(() => {
+    setValue(query);
+    setInputValue(query);
+  }, [query]);
```
Comment (ericbakenhus) [inline src/components/events-discovery/components/SearchBar/SearchBar.jsx:30]: Adjust the input when the query changes elsewhere.

File: src/components/events-discovery/components/SearchBar/SearchBar.jsx [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+        setOptions(newOptions);
+      } catch (err) {
+        setOptions([]);
+      }
+    };
+
+    fetchOptions();
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [debouncedInputValue]);
```
Comment (ericbakenhus) [inline src/components/events-discovery/components/SearchBar/SearchBar.jsx:58]: Grab a list of matching titles after the debounced input value changes. I'm not going to mess with creating a custom connector and Algolia's janky autocomplete library just to grab some suggestions.

File: src/components/events-discovery/components/SearchBar/SearchBar.jsx [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+      refine(inputValue);
+    },
+    [setValue, refine, inputValue]
+  );
+
+  const handleClear = useCallback(() => {
+    setInputValue('');
+    setValue('');
+    refine('');
+  }, [setInputValue, setValue, refine]);
```
Comment (ericbakenhus) [inline src/components/events-discovery/components/SearchBar/SearchBar.jsx:89]: Handle all the events.

File: src/components/events-discovery/eventsDiscovery.jsx [hunk 5]
```diff
@@ -86,7 +91,10 @@ const EventsDiscovery = () => (
             page: indexUiState.page,
             format: indexUiState.refinementList?.format,
             experience: indexUiState.refinementList?.experience,
-            subject: indexUiState.refinementList?.subject,
+            tags: indexUiState.refinementList?.generalTags,
+            itags: indexUiState.refinementList?.identityTags,
+            us: indexUiState.refinementList?.usRegion,
+            int: indexUiState.refinementList?.intRegion,
```
Comment (ericbakenhus) [inline src/components/events-discovery/eventsDiscovery.jsx:97]: Forgot to adjust these earlier.

File: src/hooks/useDebouncedValue.js [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+      setDebouncedValue(value);
+    }, delay);
+
+    return () => {
+      clearTimeout(handler);
+    };
+  }, [value, delay]);
+
+  return debouncedValue;
+};
```
Comment (ericbakenhus) [inline src/hooks/useDebouncedValue.js:17]: New debounced value hook.

File: src/components/events-discovery/components/Chips/ChipsComponent.jsx [hunk 7]
```diff
@@ -1,9 +1,10 @@
 import React, { useMemo } from 'react';
-import { useCurrentRefinements } from 'react-instantsearch';
+import { useCurrentRefinements, useSearchBox } from 'react-instantsearch';
 import { DateTime } from 'luxon';
 import { Chip } from './Chip';
 
 export const ChipsComponent = () => {
+  const { query, refine } = useSearchBox();
```
Comment (sherakama) [inline src/components/events-discovery/components/Chips/ChipsComponent.jsx:7]: Ah, we are getting a search box. I'll have to undo my undo in the location work.

File: src/components/events-discovery/components/SearchBar/SearchBar.jsx (hunk 3 above)
Comment (sherakama) [reply src/components/events-discovery/components/SearchBar/SearchBar.jsx:58]: The autocomplete library is tough to hammer into React but the behaviour can differ from searching for titles. I don't think our client or end users will recognize the difference but the suggestions index runs differently than the results index. For this interface, the title of the event as the 'suggestion' makes more sense IMHO.

if you're looking for a version I've managed to wrangle the autocomplete library into a working component. https://github.com/SU-SWS/eirh-subtheme/blob/1.x/src/eventshub/src/component/Autocomplete/Autocomplete.tsx

Comment (ericbakenhus) [reply src/components/events-discovery/components/SearchBar/SearchBar.jsx:58]: @sherakama The client is checking to see if they care about anything other than exact title matches. If they do, we can definitely create a suggestions index and ping that instead (or maybe get away with just `similarQuery`?). Even then, I'd probably try sticking with the base library to pull in hits. Interacting with the autocomplete library makes me itchy. I really hate having to override everything with quintuply-nested parameters and templates. I guess if you've done all the work it wouldn't be too bad though... 😆

**Pull request merged**